hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ChunkInputStream.java

```diff
@@ -290,13 +293,27 @@ protected synchronized void releaseClient() {
     }
   }
 
+  /**
+   * Updates DatanodeBlockId which based on blockId.
+   */
+  private void updateDatanodeBlockId() throws IOException {
+    DatanodeDetails closestNode = pipelineSupplier.get().getClosestNode();
+    int replicaIdx = pipelineSupplier.get().getReplicaIndex(closestNode);
+    ContainerProtos.DatanodeBlockID.Builder builder = blockID.getDatanodeBlockIDProtobufBuilder();
+    if (replicaIdx > 0) {
+      builder.setReplicaIndex(replicaIdx);
+    }
+    datanodeBlockID = builder.build();
+  }
+
   /**
    * Acquire new client if previous one was released.
    */
   protected synchronized void acquireClient() throws IOException {
     if (xceiverClientFactory != null && xceiverClient == null) {
       xceiverClient = xceiverClientFactory.acquireClientForReadData(
           pipelineSupplier.get());
+      updateDatanodeBlockId();
```

Comment:
Please store `Pipeline pipeline = pipelineSupplier.get()` and pass it to `updateDatanodeBlockId()`.

Reply:
done

Reply:
There is one more `pipelineSupplier.get()` leftover in `updateDatanodeBlockId()`:

https://github.com/apache/ozone/blob/1476d596119dada3fbc1ab1ea247d3bf54cc24bd/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ChunkInputStream.java#L301